File: packages/nodejs/packages/bridge/src/envelope.ts
import type { BleAdvertisement, DeviceReading } from "@btw/core";
import { z } from "zod";

export const payloadKindSchema = z.enum([
  "temperature_c",
  "humidity_pct",
  "switch",
  "binary_sensor",
  "raw_bytes",
  "generic",
]);

export const deviceEnvelopeSchema = z.object({
  schema_version: z.literal(1),
  source: z.object({
    runtime: z.enum(["python", "nodejs", "esp32"]),
    bridge_id: z.string().nullable().optional(),
    adapter: z.string().nullable().optional(),
  }),
  device: z.object({
    address: z.string(),
    name: z.string().nullable().optional(),
    manufacturer_data_hex: z.string().nullable().optional(),
    service_uuids: z.array(z.string()),
  }),
  payload: z.object({
    kind: payloadKindSchema,
    data: z.record(z.unknown()),
    characteristic_uuid: z.string().nullable().optional(),
  }),
  timestamp_ms: z.number().int().nonnegative(),
  rssi: z.number().int().nullable().optional(),
});

export type DeviceEnvelope = z.infer<typeof deviceEnvelopeSchema>;

export function envelopeFromReading(
  adv: BleAdvertisement,
  reading: DeviceReading,
  meta: { bridgeId?: string | null; adapter?: string | null },
): DeviceEnvelope {
  let manufacturerHex: string | null = null;
  if (adv.manufacturerData.size) {
    const parts = [...adv.manufacturerData.entries()].sort((a, b) => a[0] - b[0]);
    manufacturerHex = Buffer.concat(parts.map(([, b]) => b)).toString("hex");
  }
  return {
    schema_version: 1,
    source: {
      runtime: "nodejs",
      bridge_id: meta.bridgeId ?? null,
      adapter: meta.adapter ?? null,
    },
    device: {
      address: adv.address,
      name: adv.name,
      manufacturer_data_hex: manufacturerHex,
      service_uuids: adv.serviceUuids,
    },
    payload: {
      kind: reading.kind as DeviceEnvelope["payload"]["kind"],
      data: reading.data,
      characteristic_uuid: reading.characteristicUuid ?? null,
    },
    timestamp_ms: Date.now(),
    rssi: adv.rssi,
  };
}

File: packages/python/btw_bridge/src/btw_bridge/envelope.py
from __future__ import annotations

import time
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from btw_core.models import BleAdvertisement, DeviceReading, PayloadKind


class SourceMeta(BaseModel):
    model_config = ConfigDict(frozen=True)

    runtime: Literal["python", "nodejs", "esp32"]
    bridge_id: str | None = None
    adapter: str | None = None


class DeviceMeta(BaseModel):
    model_config = ConfigDict(frozen=True)

    address: str
    name: str | None = None
    manufacturer_data_hex: str | None = None
    service_uuids: list[str] = Field(default_factory=list)


class PayloadModel(BaseModel):
    model_config = ConfigDict(frozen=True, use_enum_values=True)

    kind: PayloadKind
    data: dict[str, Any]
    characteristic_uuid: str | None = None


class DeviceEnvelope(BaseModel):
    model_config = ConfigDict(frozen=True)

    schema_version: Literal[1] = 1
    source: SourceMeta
    device: DeviceMeta
    payload: PayloadModel
    timestamp_ms: int
    rssi: int | None = None

    def model_dump_json_bytes(self) -> bytes:
        return self.model_dump_json().encode("utf-8")


class EnvelopeFactory:
    def __init__(
        self, *, runtime: Literal["python", "nodejs", "esp32"], bridge_id: str | None
    ) -> None:
        self._runtime = runtime
        self._bridge_id = bridge_id

    def from_reading(
        self,
        adv: BleAdvertisement,
        reading: DeviceReading,
        *,
        adapter: str | None = None,
    ) -> DeviceEnvelope:
        md_hex: str | None = None
        if adv.manufacturer_data:
            blob = b"".join(v for _, v in sorted(adv.manufacturer_data.items()))
            md_hex = blob.hex() if blob else None
        return DeviceEnvelope(
            source=SourceMeta(
                runtime=self._runtime, bridge_id=self._bridge_id, adapter=adapter
            ),
            device=DeviceMeta(
                address=adv.address,
                name=adv.name,
                manufacturer_data_hex=md_hex,
                service_uuids=list(adv.service_uuids),
            ),
            payload=PayloadModel(
                kind=reading.kind,
                data=reading.data,
                characteristic_uuid=reading.characteristic_uuid,
            ),
            timestamp_ms=int(time.time() * 1000),
            rssi=adv.rssi,
        )
